## CPSC 390 Fall 24 Project 1 - Search
Name: Dylan Ernst

### **SOURCES**
- Code references/outline were adapted from:

### Part 1 Sources 
  - https://stackoverflow.com/questions/8922060/how-to-trace-the-path-in-a-breadth-first-search
  - https://www.geeksforgeeks.org/dsa/breadth-first-search-or-bfs-for-a-graph/
  - https://www.geeksforgeeks.org/dsa/depth-first-search-or-dfs-for-a-graph/
  - https://stackoverflow.com/questions/43430309/depth-first-search-dfs-code-in-python

### Part 2 Sources
  - https://www.geeksforgeeks.org/artificial-intelligence/a-algorithm-and-its-heuristic-search-strategy-in-artificial-intelligence/


### AI Usage Notice
ChatGPT was used only for minor syntax clarification and for improving wording in comments and explanations. Algorithm structure was informed by the cited sources, and all implementation and problem specific decisions were completed independently.


### The following code snippet uses a dictionary of dictionaries to encode the Romania roadmap and demonstrates how to get the elements.

In [10]:
roadmap = dict(
    Arad=dict(Zerind=75, Sibiu=140, Timisoara=118),
    Bucharest=dict(Urziceni=85, Pitesti=101, Giurgiu=90, Fagaras=211),
    Craiova=dict(Drobeta=120, Rimnicu=146, Pitesti=138),
    Drobeta=dict(Craiova=120, Mehadia=75),
    Eforie=dict(Hirsova=86),
    Fagaras=dict(Bucharest=211, Sibiu=99),
    Giurgiu=dict(Bucharest=90),
    Hirsova=dict(Eforie=86, Urziceni=98),
    Iasi=dict(Vaslui=92, Neamt=87),
    Lugoj=dict(Timisoara=111, Mehadia=70),
    Mehadia=dict(Lugoj=70, Drobeta=75),
    Neamt=dict(Iasi=87),
    Oradea=dict(Zerind=71, Sibiu=151),
    Pitesti=dict(Bucharest=101, Rimnicu=97, Craiova=138),
    Rimnicu=dict(Craiova=146, Sibiu=80, Pitesti=97),
    Sibiu=dict(Arad=140, Fagaras=99, Oradea=151, Rimnicu=80),
    Timisoara=dict(Arad=118,Lugoj=111),
    Vaslui=dict(Iasi=92, Urziceni=98),
    Urziceni=dict(Vaslui=142, Bucharest=85, Hirsova=98),
    Zerind=dict(Arad=75, Oradea=71))

Demonstration of how to use this data structure of roadmap by printing the map elements and performing a sanity check -

Please remove or comment out this section in your final submission

In [11]:
# for city, neighbors in roadmap.items():
#   print(city, neighbors)
#   for next_city, distance in neighbors.items():
#     print("  The distance from %s to %s is %d" % (city, next_city, distance))
#     if next_city not in roadmap:
#       print(" ----- ERROR! %s is not in the city node list -----" % next_city)

**Your code starts below.  Please organized them in multiple cells with a formatted text cell before each code cell to explain the purpose of the code cell**

### Breadth-First Search (BFS) — Arad to Bucharest
BFS explores the graph level by level using a **queue (FIFO)**. At each step the expanded node and the current frontier are printed in the required `Parent_city->Current_city(Path_cost)` format. When the goal is found the solution path and total cost are printed.

In [ ]:
from collections import deque

# 
def BFS(graph, start):
    # Create a set to keep track of visited nodes
    # Using a set prevents duplicates and allows for O(1) lookup
    visited = set()

    # Create a double ended queue(FIFO O(1) operations) to keep track of nodes to visit
    queue = deque([start]) 
    
    # add the first node to the visited set
    visited.add(start)

    # parent dictionary to keep track of the path
    parent = {start: None}

    # cost dictionary to keep track of the cost
    cost = {start: 0}

    # Begin BFS loop
    while queue:
        node = queue.popleft()  # get the first node from the queue

        # Print expanded node in required format “Parent_city->Current_city(Path_cost)”
        if parent[node] is None:
            expanded_str = f"START->{node}({cost[node]})"
        else:
            expanded_str = f"{parent[node]}->{node}({cost[node]})"
        print("Expanded:", expanded_str)

        # Check if goal is reached
        if node == "Bucharest":
            break

        # Iterate through neighbors 
        for neighbor, weight in graph[node].items():
            if neighbor not in visited:
                visited.add(neighbor)  # mark as visited
                parent[neighbor] = node  # track parent for path reconstruction
                cost[neighbor] = cost[node] + weight  # cumulative cost from start
                queue.append(neighbor)  # add neighbor to queue

        # Print frontier after expansion
        frontier_strs = []
        for n in queue:
            if parent[n] is None:
                frontier_strs.append(f"START->{n}({cost[n]})")
            else:
                frontier_strs.append(f"{parent[n]}->{n}({cost[n]})")
        print("Frontier:", frontier_strs)

    # Backtrack through the parent dictionary to reconstruct the path
    # The parent directory is the first parent that discovered the node
    target = "Bucharest"
    if target in parent:
        path = []
        cur = target
        while cur is not None:
            path.append(cur)
            cur = parent[cur]
        path.reverse()

        print("\nSolution path:", " -> ".join(path))
        print("Total path cost:", cost[target])

    return visited


BFS(roadmap, "Arad")

Expanded: START->Arad(0)
Frontier: ['Arad->Zerind(75)', 'Arad->Sibiu(140)', 'Arad->Timisoara(118)']
Expanded: Arad->Zerind(75)
Frontier: ['Arad->Sibiu(140)', 'Arad->Timisoara(118)', 'Zerind->Oradea(146)']
Expanded: Arad->Sibiu(140)
Frontier: ['Arad->Timisoara(118)', 'Zerind->Oradea(146)', 'Sibiu->Fagaras(239)', 'Sibiu->Rimnicu(220)']
Expanded: Arad->Timisoara(118)
Frontier: ['Zerind->Oradea(146)', 'Sibiu->Fagaras(239)', 'Sibiu->Rimnicu(220)', 'Timisoara->Lugoj(229)']
Expanded: Zerind->Oradea(146)
Frontier: ['Sibiu->Fagaras(239)', 'Sibiu->Rimnicu(220)', 'Timisoara->Lugoj(229)']
Expanded: Sibiu->Fagaras(239)
Frontier: ['Sibiu->Rimnicu(220)', 'Timisoara->Lugoj(229)', 'Fagaras->Bucharest(450)']
Expanded: Sibiu->Rimnicu(220)
Frontier: ['Timisoara->Lugoj(229)', 'Fagaras->Bucharest(450)', 'Rimnicu->Craiova(366)', 'Rimnicu->Pitesti(317)']
Expanded: Timisoara->Lugoj(229)
Frontier: ['Fagaras->Bucharest(450)', 'Rimnicu->Craiova(366)', 'Rimnicu->Pitesti(317)', 'Lugoj->Mehadia(299)']
Expanded: Faga

{'Arad',
 'Bucharest',
 'Craiova',
 'Fagaras',
 'Lugoj',
 'Mehadia',
 'Oradea',
 'Pitesti',
 'Rimnicu',
 'Sibiu',
 'Timisoara',
 'Zerind'}

### Depth-First Search (DFS) — Arad to Bucharest
DFS explores the graph by going as deep as possible along each branch before backtracking. It uses a **stack (LIFO)** instead of a queue. At each step the expanded node and the current frontier are printed in the required `Parent_city->Current_city(Path_cost)` format. When the goal is found the solution path and total cost are printed.

In [13]:
def DFS(graph, start):
    # Create a set to keep track of visited nodes
    # Using a set prevents duplicates and allows for O(1) lookup
    visited = set()

    # Use a list as a stack (LIFO) for DFS frontier
    # Each entry is a tuple: (city, parent_city, cumulative_cost)
    stack = [(start, None, 0)]

    # parent dictionary to reconstruct the path once the goal is found
    parent = {}

    # cost dictionary to track cumulative path cost to each node
    cost = {}

    # Begin DFS loop
    while stack:
        node, par, node_cost = stack.pop()  # pop from top of stack (LIFO)

        # Skip nodes that have already been visited
        if node in visited:
            continue

        # Mark node as visited and record its parent and cost
        visited.add(node)
        parent[node] = par
        cost[node] = node_cost

        # Print expanded node in required format "Parent_city->Current_city(Path_cost)"
        if par is None:
            expanded_str = f"START->{node}({node_cost})"
        else:
            expanded_str = f"{par}->{node}({node_cost})"
        print("Expanded:", expanded_str)

        # Check if goal is reached
        if node == "Bucharest":
            break

        # Push neighbors onto the stack in reverse order so that the first
        # neighbor in the adjacency list is processed first (sits on top of stack)
        for neighbor, weight in reversed(list(graph[node].items())):
            if neighbor not in visited:
                stack.append((neighbor, node, node_cost + weight))

        # Print frontier (stack) after expansion — show top-of-stack first
        frontier_strs = []
        for n, p, c in reversed(stack):
            if n not in visited:
                if p is None:
                    frontier_strs.append(f"START->{n}({c})")
                else:
                    frontier_strs.append(f"{p}->{n}({c})")
        print("Frontier:", frontier_strs)

    # Backtrack through the parent dictionary to reconstruct the solution path
    target = "Bucharest"
    if target in parent:
        path = []
        cur = target
        while cur is not None:
            path.append(cur)
            cur = parent[cur]
        path.reverse()

        print("\nSolution path:", " -> ".join(path))
        print("Total path cost:", cost[target])

    return visited


DFS(roadmap, "Arad")

Expanded: START->Arad(0)
Frontier: ['Arad->Zerind(75)', 'Arad->Sibiu(140)', 'Arad->Timisoara(118)']
Expanded: Arad->Zerind(75)
Frontier: ['Zerind->Oradea(146)', 'Arad->Sibiu(140)', 'Arad->Timisoara(118)']
Expanded: Zerind->Oradea(146)
Frontier: ['Oradea->Sibiu(297)', 'Arad->Sibiu(140)', 'Arad->Timisoara(118)']
Expanded: Oradea->Sibiu(297)
Frontier: ['Sibiu->Fagaras(396)', 'Sibiu->Rimnicu(377)', 'Arad->Timisoara(118)']
Expanded: Sibiu->Fagaras(396)
Frontier: ['Fagaras->Bucharest(607)', 'Sibiu->Rimnicu(377)', 'Arad->Timisoara(118)']
Expanded: Fagaras->Bucharest(607)

Solution path: Arad -> Zerind -> Oradea -> Sibiu -> Fagaras -> Bucharest
Total path cost: 607


{'Arad', 'Bucharest', 'Fagaras', 'Oradea', 'Sibiu', 'Zerind'}

### A* Search — Arad to Bucharest
A* search uses a priority queue (min-heap) ordered by f(n) = g(n) + h(n), where g(n) is the path cost from the start and h(n) is the straight-line distance heuristic to Bucharest. At each step the expanded node and the current frontier are printed in the format `Parent_city->Current_city(Path_cost, Estimated_total_cost)`. When the goal is found the solution path and total cost are printed.

In [14]:
import heapq

# Straight-line distance heuristic to Bucharest (from Fig. 2)
sld_to_bucharest = {
    "Arad": 366,
    "Bucharest": 0,
    "Craiova": 160,
    "Drobeta": 242,
    "Eforie": 161,
    "Fagaras": 176,
    "Giurgiu": 77,
    "Hirsova": 151,
    "Iasi": 226,
    "Lugoj": 244,
    "Mehadia": 241,
    "Neamt": 234,
    "Oradea": 380,
    "Pitesti": 100,
    "Rimnicu": 193,
    "Sibiu": 253,
    "Timisoara": 329,
    "Urziceni": 80,
    "Vaslui": 199,
    "Zerind": 374,
}

def heuristic(city, goal):
    # Return the straight-line distance from city to the goal (Bucharest)
    return sld_to_bucharest[city]

def a_star(graph, start, goal):
    # Priority queue entries: (f_score, city)
    open_set = []
    heapq.heappush(open_set, (heuristic(start, goal), start))

    # Dictionary to track the parent of each node for path reconstruction
    came_from = {}

    # g_score: actual cost from start to each node (default infinity)
    g_score = {node: float('inf') for node in graph}
    g_score[start] = 0

    # f_score: estimated total cost through each node (default infinity)
    f_score = {node: float('inf') for node in graph}
    f_score[start] = heuristic(start, goal)

    # Set of already-expanded (closed) nodes
    closed = set()

    # Begin A* loop
    while open_set:
        _, current = heapq.heappop(open_set)  # pop node with lowest f_score

        # Skip if already expanded
        if current in closed:
            continue
        closed.add(current)

        # Determine parent for printing
        if current in came_from:
            parent_city = came_from[current]
        else:
            parent_city = "START"

        # Print expanded node: "Parent_city->Current_city(Path_cost, Estimated_total_cost)"
        print(f"Expanded: {parent_city}->{current}({g_score[current]}, {f_score[current]})")

        # Check if goal is reached
        if current == goal:
            # Reconstruct the solution path
            path = [current]
            node = current
            while node in came_from:
                node = came_from[node]
                path.append(node)
            path.reverse()

            print("\nSolution path:", " -> ".join(path))
            print("Total path cost:", g_score[goal])
            return path, g_score[goal]

        # Explore neighbors
        for neighbor, cost in graph[current].items():
            if neighbor in closed:
                continue
            tentative_g = g_score[current] + cost
            if tentative_g < g_score[neighbor]:
                # Found a better path to this neighbor
                came_from[neighbor] = current
                g_score[neighbor] = tentative_g
                f_score[neighbor] = tentative_g + heuristic(neighbor, goal)
                heapq.heappush(open_set, (f_score[neighbor], neighbor))

        # Print frontier after expansion
        # Filter out already-closed nodes
        frontier_items = [(f, n) for f, n in open_set if n not in closed]
        frontier_items.sort()
        frontier_strs = []
        for f, n in frontier_items:
            if n in came_from:
                frontier_strs.append(f"{came_from[n]}->{n}({g_score[n]}, {f_score[n]})")
            else:
                frontier_strs.append(f"START->{n}({g_score[n]}, {f_score[n]})")
        print("Frontier:", frontier_strs)

    print("No path found.")
    return None, None


a_star(roadmap, "Arad", "Bucharest")

Expanded: START->Arad(0, 366)
Frontier: ['Arad->Sibiu(140, 393)', 'Arad->Timisoara(118, 447)', 'Arad->Zerind(75, 449)']
Expanded: Arad->Sibiu(140, 393)
Frontier: ['Sibiu->Rimnicu(220, 413)', 'Sibiu->Fagaras(239, 415)', 'Arad->Timisoara(118, 447)', 'Arad->Zerind(75, 449)', 'Sibiu->Oradea(291, 671)']
Expanded: Sibiu->Rimnicu(220, 413)
Frontier: ['Sibiu->Fagaras(239, 415)', 'Rimnicu->Pitesti(317, 417)', 'Arad->Timisoara(118, 447)', 'Arad->Zerind(75, 449)', 'Rimnicu->Craiova(366, 526)', 'Sibiu->Oradea(291, 671)']
Expanded: Sibiu->Fagaras(239, 415)
Frontier: ['Rimnicu->Pitesti(317, 417)', 'Arad->Timisoara(118, 447)', 'Arad->Zerind(75, 449)', 'Fagaras->Bucharest(450, 450)', 'Rimnicu->Craiova(366, 526)', 'Sibiu->Oradea(291, 671)']
Expanded: Rimnicu->Pitesti(317, 417)
Frontier: ['Pitesti->Bucharest(418, 418)', 'Arad->Timisoara(118, 447)', 'Arad->Zerind(75, 449)', 'Pitesti->Bucharest(418, 418)', 'Rimnicu->Craiova(366, 526)', 'Sibiu->Oradea(291, 671)']
Expanded: Pitesti->Bucharest(418, 418)

Sol

(['Arad', 'Sibiu', 'Rimnicu', 'Pitesti', 'Bucharest'], 418)

## **Extra Credit — Why A* Finds the Cheapest Route**

A* keeps a scoreboard for every city. The score is **f(n) = g(n) + h(n)** where g(n) is how much road distance we have already paid to reach the city and h(n) is a guess of the distance still left. We use the straight-line distance to Bucharest for h(n). A straight line is never longer than a real road, so this guess never claims the trip will cost less than it really does. That makes it an **admissible** heuristic.

Now think about running A* from Arad. Suppose it just pulled Bucharest off the priority queue after following the longer route Arad → Sibiu → Fagaras → Bucharest (cost 450). There is still another city waiting in the queue—Pitesti—because Pitesti lies on the cheaper route Arad → Sibiu → Rimnicu → Pitesti → Bucharest. Pitesti’s score is **f(Pitesti) = 317 (actual cost so far) + 100 (straight-line guess) = 417**, which is smaller than 450. Because A* always takes the city with the smallest score, it must remove Pitesti before removing that expensive Bucharest node. Expanding Pitesti reveals Bucharest with a total cost of 418, and that goal is accepted first.

That little story is exactly what the proof in the textbook is saying: if the heuristic never overestimates, then any unfinished piece of the optimal path will always look at least as good as every suboptimal goal. So A* cannot finish an expensive route before it has finished the cheap one. In this map, that means A* always ends with the 418‑mile path Arad → Sibiu → Rimnicu → Pitesti → Bucharest.